### PR TITLE
Add the `array_histogram` operator

### PR DESCRIPTION
**Pull request by nathanwilk7**
## Description

Adds the `array_histogram` function along with tests and documentation for it.

Fixes https://github.com/trinodb/trino/issues/14725

## Additional context and related issues

The `array_histogram` function was requested in [issue 14725](https://github.com/trinodb/trino/issues/14725). `martint` suggested the name `array_histogram` to make it consistent with the existing `histogram` function. I've implemented this as a scalar function with the behavior shown below (the input to the function is an array and the output is a map, nulls are ignored).

```sql
SELECT array_histogram(a) FROM (VALUES (ARRAY[1, 2, 1]), (ARRAY[42, 7, 42, null])) t(a);
    _col0
-------------
 {1=2, 2=1}
 {42=2, 7=1}
```

This is my first attempt to contribute to Trino (got the CLA confirmation email and org invite yesterday), feedback is very welcome and I will happily address it, I'm eager to understand the codebase more (this PR was essentially just pattern matching, so I expect there are things I've misused or misunderstood). I've left some comments with questions for reviewers inline as well.

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# General
* Add the `array_histogram` function. ({issue}`14725 `)
```

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/operator/scalar/ArrayHistogram.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+                    argumentTypes = "T",
+                    convention = @Convention(arguments = BLOCK_POSITION, result = FAIL_ON_NULL))
+            BlockTypeOperators.BlockPositionHashCode hashCodeOperator,
+            @TypeParameter("map(T, BIGINT)") MapType mapType,
+            @SqlType("array(T)") Block arrayBlock)
+    {
+        TypedHistogram histogram = new SingleTypedHistogram(elementType, equalOperator, hashCodeOperator, 10);
+        int positionCount = arrayBlock.getPositionCount();
+        for (int position = 0; position < positionCount; position++) {
+            if (!arrayBlock.isNull(position)) {
```
Comment (nathanwilk7) [inline core/trino-main/src/main/java/io/trino/operator/scalar/ArrayHistogram.java:62]: Do you agree with my choice to ignore nulls? The alternatives would be to either include nulls as a key in the output map with their respective count (assuming null can actually be a key in a map) or to return null as the result of the function if any element in the array is null.

File: core/trino-main/src/test/java/io/trino/operator/scalar/TestArrayHistogram.java [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    @Test
+    public void testLargeArray()
+    {
+        assertThat(assertions.expression("array_histogram(a)")
+                             .binding("a", "ARRAY[42, 42, 42, 7, 1, 7, 1, 42, 7, 1, 1, 42, 7, 42, 1, 7, 2, 3, 7, 42, 42]"))
+                .isEqualTo(Map.of(1, 5L, 2, 1L, 3, 1L, 7, 6L, 42, 8L));
+    }
+
+    @Test
+    public void testTypes()
```
Comment (nathanwilk7) [inline core/trino-main/src/test/java/io/trino/operator/scalar/TestArrayHistogram.java:106]: This test is probably overkill, I've tested many of the different types available in Trino, but we could probably strip this down to only a few types. Let me know if you want it kept as is or if you want types cut from the test. The test takes just over 1 second to run on my box.

File: docs/src/main/sphinx/functions/array.rst [hunk 3]
```diff
@@ -60,6 +60,10 @@ Array functions
 
     Returns an array of elements in ``x`` but not in ``y``, without duplicates.
 
+.. function:: array_histogram(x) -> map(K, bigint)
```
Comment (nathanwilk7) [inline docs/src/main/sphinx/functions/array.rst:63]: This return type of `map(K, bigint)` is not rendering for me in safari or chrome. The same issue happens for the existing `histogram` function from the [Aggregate functions page](https://trino.io/docs/current/functions/aggregate.html#map-aggregate-functions) (shown below in a screenshot). I will look into this rendering issue separate from this PR.
image

**Comment by nathanwilk7** [comment]
@bitsondatadev @findinpath Would either of you be interested in reviewing this? Or if not, would you be able to suggest someone who could do so?

**Review by nathanwilk7 — COMMENTED** [review]
Leaving questions for reviewers

File: core/trino-main/src/main/java/io/trino/operator/scalar/ArrayHistogram.java (hunk 1 above)
Comment (bitsondatadev) [reply core/trino-main/src/main/java/io/trino/operator/scalar/ArrayHistogram.java:62]: I'd say this is a pretty fair assumptions given that's how most of the [other aggregate operators work](https://trino.io/docs/current/functions/aggregate.html).

> Except for [count()](https://trino.io/docs/current/functions/aggregate.html#count), [count_if()](https://trino.io/docs/current/functions/aggregate.html#count_if), [max_by()](https://trino.io/docs/current/functions/aggregate.html#max_by), [min_by()](https://trino.io/docs/current/functions/aggregate.html#min_by) and [approx_distinct()](https://trino.io/docs/current/functions/aggregate.html#approx_distinct), all of these aggregate functions ignore null values and return null for no input rows or when all values are null. For example, [sum()](https://trino.io/docs/current/functions/aggregate.html#sum) returns null rather than zero and [avg()](https://trino.io/docs/current/functions/aggregate.html#avg) does not include null values in the count. The coalesce function can be used to convert null into zero.

File: core/trino-main/src/test/java/io/trino/operator/scalar/TestArrayHistogram.java (hunk 2 above)
Comment (bitsondatadev) [reply core/trino-main/src/test/java/io/trino/operator/scalar/TestArrayHistogram.java:106]: IMO the more the merrier :). This is a good representation of what's possible and a good test to glean examples from.

File: docs/src/main/sphinx/functions/array.rst (hunk 3 above)
Comment (bitsondatadev) [reply docs/src/main/sphinx/functions/array.rst:63]: This is likely something me and @mosabua need to look into. Please create another issue around this and tag us in it please!

File: core/trino-main/src/test/java/io/trino/operator/scalar/TestArrayHistogram.java [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    {
+        assertThat(assertions.expression("array_histogram(a)")
+                .binding("a", "ARRAY[null]"))
+                .isEqualTo(Map.of());
+
+        assertThat(assertions.expression("array_histogram(a)")
+                .binding("a", "ARRAY[42, null]"))
+                .isEqualTo(Map.of(42, 1L));
+    }
+
```
Comment (bitsondatadev) [inline core/trino-main/src/test/java/io/trino/operator/scalar/TestArrayHistogram.java:96]: Perhaps we should test negatives, zeros, and other odd values.

File: core/trino-main/src/main/java/io/trino/operator/scalar/ArrayHistogram.java (hunk 1 above)
Comment (nathanwilk7) [reply core/trino-main/src/main/java/io/trino/operator/scalar/ArrayHistogram.java:62]: Just found the code which does not allow nulls as a key in maps in Trino: https://github.com/trinodb/trino/blob/19f61fc56984bd98c617d8e72214a61237f8b256/core/trino-main/src/main/java/io/trino/operator/scalar/MapConstructor.java#L135

**Comment by nathanwilk7** [comment]
@bitsondatadev Per your request, I added some more edge case tests like negative numbers, zero, IP addresses, and timestamps with time zones.

**Comment by mosabua** [comment]
Given that this is a new function and hence part of the language, this will need a review from @martint

**Comment by findinpath** [comment]
Please add "Fixes #14725" to the description of the ticket in order to eventually close the initial issue once this PR lands.

File: core/trino-main/src/test/java/io/trino/operator/scalar/TestArrayHistogram.java [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+        assertThat(assertions.expression("array_histogram(a)")
+                .binding("a", "ARRAY[]"))
+                .isEqualTo(Map.of());
+    }
+
+    @Test
+    public void testNullsIgnored()
+    {
+        assertThat(assertions.expression("array_histogram(a)")
+                .binding("a", "ARRAY[null]"))
```
Comment (findinpath) [inline core/trino-main/src/test/java/io/trino/operator/scalar/TestArrayHistogram.java:89]: nit: Use `NULL` in SQL code instead of `null`.

File: core/trino-main/src/test/java/io/trino/operator/scalar/TestArrayHistogram.java [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+    {
+        assertThat(assertions.expression("array_histogram(a)")
+                .binding("a", "ARRAY[]"))
+                .isEqualTo(Map.of());
+    }
+
+    @Test
+    public void testNullsIgnored()
+    {
+        assertThat(assertions.expression("array_histogram(a)")
```
Comment (findinpath) [inline core/trino-main/src/test/java/io/trino/operator/scalar/TestArrayHistogram.java:88]: add an extra assertions with an array containing multiple `NULL` elements.

File: core/trino-main/src/main/java/io/trino/operator/scalar/ArrayHistogram.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+import io.trino.spi.type.Type;
+import io.trino.type.BlockTypeOperators;
+
+import static io.trino.spi.function.InvocationConvention.InvocationArgumentConvention.BLOCK_POSITION;
+import static io.trino.spi.function.InvocationConvention.InvocationReturnConvention.FAIL_ON_NULL;
+import static io.trino.spi.function.InvocationConvention.InvocationReturnConvention.NULLABLE_RETURN;
+
+@Description("Return a map containing the counts of the elements in the array")
+@ScalarFunction(value = "array_histogram")
+public final class ArrayHistogram
```
Comment (findinpath) [inline core/trino-main/src/main/java/io/trino/operator/scalar/ArrayHistogram.java:38]: Change class name to `ArrayHistogramFunction`.

File: core/trino-main/src/main/java/io/trino/operator/scalar/ArrayHistogram.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+                    operator = OperatorType.EQUAL,
+                    argumentTypes = {"T", "T"},
+                    convention = @Convention(arguments = {BLOCK_POSITION, BLOCK_POSITION}, result = NULLABLE_RETURN))
+            BlockTypeOperators.BlockPositionEqual equalOperator,
+            @OperatorDependency(
+                    operator = OperatorType.HASH_CODE,
+                    argumentTypes = "T",
+                    convention = @Convention(arguments = BLOCK_POSITION, result = FAIL_ON_NULL))
+            BlockTypeOperators.BlockPositionHashCode hashCodeOperator,
+            @TypeParameter("map(T, BIGINT)") MapType mapType,
```
Comment (findinpath) [inline core/trino-main/src/main/java/io/trino/operator/scalar/ArrayHistogram.java:56]: `map(T, BIGINT)` -> `map(T, bigint)`

**Comment by nathanwilk7** [comment]
@findinpath Feedback addressed, ready for re-review at your convenience.

Let me know if I should be doing commits differently, it's not clear to me if I should be autosquashing and force pushing or just leaving the fixup commits.

**Comment by findinpath** [comment]
Please squash the commits

File: core/trino-main/src/main/java/io/trino/operator/scalar/ArrayHistogramFunction.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+            BlockTypeOperators.BlockPositionEqual equalOperator,
+            @OperatorDependency(
+                    operator = OperatorType.HASH_CODE,
+                    argumentTypes = "T",
+                    convention = @Convention(arguments = BLOCK_POSITION, result = FAIL_ON_NULL))
+            BlockTypeOperators.BlockPositionHashCode hashCodeOperator,
+            @TypeParameter("map(T, bigint)") MapType mapType,
+            @SqlType("array(T)") Block arrayBlock)
+    {
+        TypedHistogram histogram = new SingleTypedHistogram(elementType, equalOperator, hashCodeOperator, 10);
```
Comment (findinpath) [inline core/trino-main/src/main/java/io/trino/operator/scalar/ArrayHistogramFunction.java:59]: Let's use a constant for `10`.

File: core/trino-main/src/main/java/io/trino/operator/scalar/ArrayHistogramFunction.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+        int positionCount = arrayBlock.getPositionCount();
+        for (int position = 0; position < positionCount; position++) {
+            if (!arrayBlock.isNull(position)) {
+                histogram.add(position, arrayBlock, 1L);
+            }
+        }
+        BlockBuilder blockBuilder = mapType.createBlockBuilder(null, 1);
```
Comment (findinpath) [inline core/trino-main/src/main/java/io/trino/operator/scalar/ArrayHistogramFunction.java:66]: We could add `getPositionCount() {return values.getPositionCount();}` to `SingleTypedHistogram` to avoid expanding on the fly the map block builder internally during the `serialize` operation.

File: core/trino-main/src/test/java/io/trino/operator/scalar/TestArrayHistogramFunction.java [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+                             .binding("a", "ARRAY[IPADDRESS '10.0.0.1', IPADDRESS '::ffff:a00:1']"))
+                .matches("MAP(ARRAY[IPADDRESS '::ffff:a00:1'], CAST(ARRAY[2] AS ARRAY(BIGINT)))");
+
+        assertThat(assertions.expression("transform_keys(array_histogram(a), (k, v) -> k AT TIME ZONE 'UTC')")
+                             .binding("a", "ARRAY[TIMESTAMP '2001-01-01 01:00:00.000 UTC', TIMESTAMP '2001-01-01 02:00:00.000 +01:00']"))
+                .matches("MAP(ARRAY[TIMESTAMP '2001-01-01 01:00:00.000 UTC'], CAST(ARRAY[2] AS ARRAY(BIGINT)))");
+    }
+
+    @Test
+    public void testTypes()
```
Comment (findinpath) [inline core/trino-main/src/test/java/io/trino/operator/scalar/TestArrayHistogramFunction.java:130]: You could use `org.testng.annotations.DataProvider` to avoid repeating the same query over and over.

File: core/trino-main/src/test/java/io/trino/operator/scalar/TestArrayHistogramFunction.java [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+    }
+
+    @Test
+    public void testBasic()
+    {
+        assertThat(assertions.expression("array_histogram(a)")
+                             .binding("a", "ARRAY[42]"))
+                .isEqualTo(Map.of(42, 1L));
+
+        assertThat(assertions.expression("array_histogram(a)")
```
Comment (martint) [inline core/trino-main/src/test/java/io/trino/operator/scalar/TestArrayHistogramFunction.java:52]: Use assertions.function(), which provides a shortcut to avoid having to use binding()

File: core/trino-main/src/test/java/io/trino/operator/scalar/TestArrayHistogramFunction.java (hunk 11 above)
Comment (martint) [reply core/trino-main/src/test/java/io/trino/operator/scalar/TestArrayHistogramFunction.java:130]: No, don’t use that. It makes it a lot harder to set breakpoints and debug when there are issues.

File: core/trino-main/src/test/java/io/trino/operator/aggregation/TestTypedHistogram.java [hunk 13]
```diff
@@ -31,25 +33,24 @@
...
                 1000);
         IntStream.range(1, 2000)
                 .flatMap(i -> IntStream.iterate(i, IntUnaryOperator.identity()).limit(i))
                 .forEach(j -> BIGINT.writeLong(inputBlockBuilder, j));
 
         Block inputBlock = inputBlockBuilder.build();
-        for (int i = 0; i < inputBlock.getPositionCount(); i++) {
-            typedHistogram.add(i, inputBlock, 1);
-        }
+        addInputBlockToTypedHistogram(typedHistogram, inputBlock);
```
Comment (nathanwilk7) [inline core/trino-main/src/test/java/io/trino/operator/aggregation/TestTypedHistogram.java:53]: Just DRYing up an existing test with the new ones I added below

**Review by nathanwilk7 — COMMENTED** [review]
Notes for reviewers

**Comment by nathanwilk7** [comment]
@findinpath @martint Addressed feedback, ready for re-review at your convenience.
- `EXPECTED_HISTOGRAM_SIZE` constant for `10`.
- `getPositionCount` method to `SingleTypedHistogram` class.
- Basic tests for new `getPositionCount` method.
- Use `histogram.getPositionCount` in call to `mapType.createBlockBuilder` to avoid on the fly expansion.
- Replace `assertions.expression` with `assertions.function`.

**Review by findinpath — APPROVED** [review]

**Review by findinpath — COMMENTED** [review]
Build is red.
`TestTypedHistogram.testGetPositionCountEmpty:118 » IllegalArgument expectedSiz`

**Comment by nathanwilk7** [comment]
@findinpath Thanks fixed the build, sorry about that, I didn't test my last addition locally and I should have let the CI tests complete before I pinged you.

**Review by findinpath — APPROVED** [review]

**Comment by findinpath** [comment]
@martint  CPTAL?

**Comment by ebyhr** [comment]
Removed `syntax-needs-review` label as we already agree the function name in #14725

File: core/trino-main/src/test/java/io/trino/operator/aggregation/TestTypedHistogram.java [hunk 14]
```diff
@@ -59,4 +60,85 @@ public void testMassive()
...
+    }
+
+    private Block inputBlockFromNumbers(List<Integer> numbersToInsert, int expectedEntries)
+    {
+        BlockBuilder inputBlockBuilder = BIGINT.createBlockBuilder(null, expectedEntries);
+        for (int n : numbersToInsert) {
+            BIGINT.writeLong(inputBlockBuilder, n);
+        }
+        return inputBlockBuilder.build();
+    }
```
Comment (raunaqmorarka) [inline core/trino-main/src/test/java/io/trino/operator/aggregation/TestTypedHistogram.java:136]: You can use `BlockAssertions#createLongsBlock`

**Comment by nathanwilk7** [comment]
Thanks for the review @raunaqmorarka , I'm out of town without my computer for a few weeks but will address your comments after that time. If someone else prefers to make the changes instead of waiting, feel free to do so!

**Comment by colebow** [comment]
> Thanks for the review @raunaqmorarka , I'm out of town without my computer for a few weeks but will address your comments after that time. If someone else prefers to make the changes instead of waiting, feel free to do so!

This should work fine! If the stale bot shows back up, we just need to make sure we remove the 'stale' label so it doesn't get auto-closed on us.

File: core/trino-main/src/main/java/io/trino/operator/scalar/ArrayHistogramFunction.java [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+                equalOperator,
+                hashCodeOperator,
+                EXPECTED_HISTOGRAM_SIZE);
+        int positionCount = arrayBlock.getPositionCount();
+        for (int position = 0; position < positionCount; position++) {
+            if (!arrayBlock.isNull(position)) {
+                histogram.add(position, arrayBlock, 1L);
+            }
+        }
+        BlockBuilder blockBuilder = mapType.createBlockBuilder(null, histogram.getPositionCount());
```
[the remark this answers is not in the document]
Comment (nathanwilk7) [reply core/trino-main/src/main/java/io/trino/operator/scalar/ArrayHistogramFunction.java:71]: Good question. I'll explain my thought process and please let me know anywhere I'm misunderstanding.

My thought was that `expectedEntries` here refers to the number of elements which we expect to insert into the `blockBuilder` being created on this line. Within the `serialize` function on the next line (72), we write each value into the `blockBuilder`. I believe the `getPositionCount` function on the histogram returns the number of keys in the histogram. So given this, my thought was that I can improve the performance of the `blockBuilder` since on line 71 I expect that `getPositionCount` entries will be inserted into the `blockBuilder` in the `serialize` call on line 72.

If I specified `expectedEntries` as `1`, then I expect the `blockBuilder` would need to resize multiple times if there were many keys in the histogram.

Any corrections welcome, thanks!

File: core/trino-main/src/main/java/io/trino/operator/scalar/ArrayHistogramFunction.java [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+            @OperatorDependency(
+                    operator = OperatorType.EQUAL,
+                    argumentTypes = {"T", "T"},
+                    convention = @Convention(arguments = {BLOCK_POSITION, BLOCK_POSITION}, result = NULLABLE_RETURN))
+            BlockTypeOperators.BlockPositionEqual equalOperator,
+            @OperatorDependency(
+                    operator = OperatorType.HASH_CODE,
+                    argumentTypes = "T",
+                    convention = @Convention(arguments = BLOCK_POSITION, result = FAIL_ON_NULL))
+            BlockTypeOperators.BlockPositionHashCode hashCodeOperator,
```
[the remark this answers is not in the document]
Comment (nathanwilk7) [reply core/trino-main/src/main/java/io/trino/operator/scalar/ArrayHistogramFunction.java:56]: Same as comment on `BlockPositionEqual`

File: core/trino-main/src/main/java/io/trino/operator/scalar/ArrayHistogramFunction.java [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+
+    @TypeParameter("T")
+    @SqlType("map(T, bigint)")
+    public static Block arrayHistogram(
+            @TypeParameter("T") Type elementType,
+            @OperatorDependency(
+                    operator = OperatorType.EQUAL,
+                    argumentTypes = {"T", "T"},
+                    convention = @Convention(arguments = {BLOCK_POSITION, BLOCK_POSITION}, result = NULLABLE_RETURN))
+            BlockTypeOperators.BlockPositionEqual equalOperator,
```
[the remark this answers is not in the document]
Comment (nathanwilk7) [reply core/trino-main/src/main/java/io/trino/operator/scalar/ArrayHistogramFunction.java:51]: You're welcome to ignore this if you prefer to leave the code as is rather than take the time to answer my question, but could you give an example of how one could static import `BlockPositionEqual`? I can't find any other examples of doing so by grepping the codebase and `BlockPositionEqual` is a non-static interface.

File: core/trino-main/src/test/java/io/trino/operator/aggregation/TestTypedHistogram.java (hunk 14 above)
Comment (nathanwilk7) [reply core/trino-main/src/test/java/io/trino/operator/aggregation/TestTypedHistogram.java]: Thanks for the suggestion, I removed the `inputBlockFromNumbers` helper functions and replaced them with `BlockAssertions#createLongsBlock`.

**Comment by nathanwilk7** [comment]
@raunaqmorarka This PR is ready for your re-review

File: core/trino-main/src/main/java/io/trino/operator/scalar/ArrayHistogramFunction.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/operator/scalar/ArrayHistogramFunction.java:71]: Looked into it more closely, I wasn't sure whether the expectedEntries propagated into the key and value builder, but it does. So using `histogram.getPositionCount()` is better.

File: core/trino-main/src/main/java/io/trino/operator/aggregation/histogram/SingleTypedHistogram.java [hunk 18]
```diff
@@ -173,6 +173,11 @@ public boolean isEmpty()
         return values.getPositionCount() == 0;
     }
 
+    public int getPositionCount()
```
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/operator/aggregation/histogram/SingleTypedHistogram.java:175]: this is fine

**Review by raunaqmorarka — APPROVED** [review]

**Comment by nathanwilk7** [comment]
@raunaqmorarka Made the final import fix. This should be ready to merge, thanks for the help
